Fix race conditions and improve robustness during socket I/O

julianz- opened the pull request:
Fixed race conditions to make socket I/O more resilient during connection teardown.

1. BufferedWriter's ``write()``: Added error handling to ignore common socket errors (e.g., ``ECONNRESET``, ``EPIPE``, ``ENOTCONN``, ``EBADF``) that occur when the underlying connection has been unexpectedly closed by the client or OS. This prevents a crash when attempting to write to a defunct socket.
2. BufferedWriters's ``close()``: Made idempotent, allowing safe repeated calls without raising exceptions.
3. Needed to add explicit handling of WINDOWS environments as these are seen to throw Windows specific ``WSAENOTSOCK`` errors.

Includes new unit tests to cover the idempotency and graceful handling of already closed underlying buffers.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cherrypy/cheroot/779)
<!-- Reviewable:end -->

julianz- committed:
Fix race condition and improve robustness during socket I/O

Fixes to make socket I/O more resilient during connection teardown.

1. BufferedWriter's write(): Added error handling to ignore common
   socket errors (e.g., ECONNRESET, EPIPE, ENOTCONN, EBADF) that occur
   when the underlying connection has been unexpectedly closed by the
   client or OS. This prevents a crash when attempting to write to a
   defunct socket.
2. BufferedWriters's close(): Made idempotent, allowing safe repeated
   calls without raising exceptions.
3. Needed to add explicit handling of WINDOWS environments as these are
   seen to throw Windows specific WSAENOTSOCK errors.

Includes new unit tests to cover the idempotency and graceful handling
of already closed underlying buffers.

webknjaz commented on cheroot/errors.py:
```diff
@@ -66,6 +72,7 @@ def plat_specific_errors(*errnames):
 
 
 acceptable_sock_shutdown_error_codes = {
+    errno.EBADF,
```
Can we have something like

```suggestion
_bad_fd_errors = {errno.EBADF}
if IS_WINDOWS:
    _bad_fd_errors(socket.WSAENOTSOCK)

acceptable_sock_shutdown_error_codes = {
    *_bad_fd_errors,
```

? Perhaps, with that fallback you're trying to add to `makefile`. We could just have that here.

webknjaz commented on cheroot/errors.py:
```diff
@@ -87,4 +94,8 @@ def plat_specific_errors(*errnames):
+acceptable_sock_shutdown_exceptions = (
+    BrokenPipeError,
+    ConnectionResetError,
+    _OpenSSL_SysCallError,
```
Perhaps avoid adding `None` here and make it conditional?

webknjaz commented on cheroot/makefile.py:
```diff
@@ -30,10 +52,59 @@ def _flush_unlocked(self):
                 n = self.raw.write(bytes(self._write_buf))
-            except io.BlockingIOError as e:
-                n = e.characters_written
+            except (io.BlockingIOError, OSError, SysCallError) as e:
+                # Check for a different error attribute depending
+                # on the exception type
+                if isinstance(e, io.BlockingIOError):
+                    n = e.characters_written
```
Keep it in a separate `except` block to avoid unnecessary nested blocks. It'll be easier to reason about.

webknjaz commented on cheroot/makefile.py:
```diff
@@ -30,10 +52,59 @@ def _flush_unlocked(self):
+
+        try:
+            super().close()
+        except (OSError, SysCallError) as e:
```
Should this be in a var?

julianz- replied:
i can reuse acceptable_sock_shutdown_error_codes?

webknjaz commented on docs/spelling_wordlist.txt:
this looks unnecessary

webknjaz commented on cheroot/server.py:
```diff
@@ -1186,12 +1189,25 @@ def ensure_headers_sent(self):
+        except (SysCallError, ConnectionError, OSError) as e:
+            error_code = e.errno if isinstance(e, OSError) else e.args[0]
+            if error_code in {
+                errno.ECONNRESET,
+                errno.EPIPE,
+                errno.ENOTCONN,
+                errno.EBADF,
```
We should probably get this from a var.

webknjaz commented on cheroot/server.py:
```diff
@@ -1186,12 +1189,25 @@ def ensure_headers_sent(self):
+            else:
+                self.conn.wfile.write(chunk)
+        except (SysCallError, ConnectionError, OSError) as e:
+            error_code = e.errno if isinstance(e, OSError) else e.args[0]
```
There's a number of places where this is repeated. It's worth having a helper function, I think.

julianz- commented on cheroot/makefile.py:
```diff
@@ -30,10 +52,59 @@ def _flush_unlocked(self):
                 n = self.raw.write(bytes(self._write_buf))
-            except io.BlockingIOError as e:
-                n = e.characters_written
+            except (io.BlockingIOError, OSError, SysCallError) as e:
```
Agreed!

julianz- commented on cheroot/server.py:
```diff
@@ -1186,12 +1189,25 @@ def ensure_headers_sent(self):
+                self.conn.wfile.write(EMPTY.join(buf))
+            else:
+                self.conn.wfile.write(chunk)
+        except (SysCallError, ConnectionError, OSError) as e:
```
revised

julianz- commented on cheroot/server.py:
```diff
@@ -1541,9 +1577,10 @@ def _close_kernel_socket(self):
         try:
             shutdown(socket.SHUT_RDWR)  # actually send a TCP FIN
-        except errors.acceptable_sock_shutdown_exceptions:
+        except acceptable_exceptions:  # pylint: disable=E0712
```
Revised. Yes that was silly - I think this was the aftermath of refactoring.

webknjaz commented:
Oh, it seems there's some more errors showing up on `main` after `pre-commit.ci`'s cache got invalidated. I'll see what needs to be done.

webknjaz commented on cheroot/test/test_makefile.py:
```diff
@@ -51,3 +53,27 @@ def test_bytes_written():
+    buffered_writer.close()
+    buffered_writer.close()  # Second call should be safe
+
+    assert buffered_writer.closed
```
This will make it explicit that the second call is performed on an already-closed buffer.
```suggestion
    buffered_writer.close()

    assert buffered_writer.closed

    buffered_writer.close()  # Second call should be safe
```

webknjaz commented on cheroot/errors.py:
```diff
@@ -64,15 +72,27 @@ def plat_specific_errors(*errnames):
-"""Errors that may happen during the connection close sequence.
 
+if IS_WINDOWS:
+    # Add Windows socket error code constant
```
This phrase just repeats what's already in code. Code comments are supposed to explain motivation, not duplicate what's right below.

webknjaz commented on cheroot/errors.py:
```diff
@@ -64,15 +72,27 @@ def plat_specific_errors(*errnames):
 
+if IS_WINDOWS:
+    # Add Windows socket error code constant
+    # WSAENOTSOCK: Socket operation on non-socket
```
This could go into a docstring.

julianz- replied:
this has been absorbed now

webknjaz commented on cheroot/errors.py:
```diff
@@ -64,15 +72,27 @@ def plat_specific_errors(*errnames):
+write_shutdown_errs = (OSError, _OpenSSL_SysCallError, ConnectionError)
 
 acceptable_sock_shutdown_error_codes = {
+    errno.EBADF,
```
We can just inline this so it's not detached in some separate place in the module.
```suggestion
    errno.EBADF,
    *((errno.WSAENOTSOCK,) if IS_WINDOWS else ()),
```

julianz- commented:
To get anywhere with the CI, I had to remove darglint ~= 1.8.1 from flake in .pre-commit-config.yaml because it is apparently very buggy and no longer supported. I could not get past errors it was throwing. I also had to fix docs/conf.py adding missing commas to the sphinx-doc list. I can remove these changes as they are unrelated to the PR.

webknjaz commented on .pre-commit-config.yaml:
This shouldn't be necessary if you're rebased on the latest `main`.

webknjaz commented on cheroot/errors.py:
```diff
@@ -87,4 +99,26 @@ def plat_specific_errors(*errnames):
+
+acceptable_sock_shutdown_exceptions = (
+    BrokenPipeError,  # Covers EPIPE and ESHUTDOWN
+    ConnectionResetError,  # Covers ECONNRESET)
```
```suggestion
    ConnectionResetError,  # Covers ECONNRESET
```

webknjaz commented on cheroot/errors.py:
```diff
@@ -66,13 +76,15 @@ def plat_specific_errors(*errnames):
     errno.ECONNRESET,  # corresponds to ConnectionResetError in Python 3
+    *((errno.WSAENOTSOCK,) if cheroot._compat.IS_WINDOWS else ()),
 }
 """Errors that may happen during the connection close sequence.
```
Missing it makes the render broken: https://cheroot--779.org.readthedocs.build/en/779/pkg/cheroot.errors/#cheroot.errors.acceptable_sock_shutdown_error_codes

webknjaz commented:
@julianz- 

> To get anywhere with the CI, I had to remove darglint ~= 1.8.1 from flake in .pre-commit-config.yaml because it is apparently very buggy and no longer supported. I could not get past errors it was throwing. I also had to fix docs/conf.py adding missing commas to the sphinx-doc list. I can remove these changes as they are unrelated to the PR.

pre-commit is passing on `main`. You might've been rebased on a temporarily broken state at some point but it shouldn't be the case anymore. So changes outside of your patch shouldn't be needed.

We can't replace `darglint` with Ruff because there's no replacement in it (https://github.com/astral-sh/ruff/issues/458). However, I thought it'd be a good idea to try migrating to https://pypi.org/p/pydoclint. Feel free to look into it separately.

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,44 @@ def _flush_unlocked(self):
+        and ignored, as these indicate a normal, expected connection teardown.
+        Other exceptions are re-raised.
+        """
+        if self.closed:  # pylint: disable=W0125
```
Use human-readable codes for pylint. A letter and a bunch of numbers don't really give away any context regarding when this can be dropped or why it's here is the first place.

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,44 @@ def _flush_unlocked(self):
+        where the underlying socket may have already been closed by the remote
+        client or another thread.
+
+        A SysCallError or OSError with ``EBADF`` or ``ENOTCONN`` is caught
```
All these are referencible
```suggestion
        A :exc:`~OpenSSL.SSL.SysCallError` or :exc:`OSError` with
        :data:`~errno.EBADF` or :data:`~errno.ENOTCONN` is caught
```

webknjaz commented on cheroot/errors.py:
```diff
@@ -87,4 +99,26 @@ def plat_specific_errors(*errnames):
 
-acceptable_sock_shutdown_exceptions = (BrokenPipeError, ConnectionResetError)
+
+def _is_acceptable_socket_shutdown_error(sock_err):
```
I'd rather have this helper only handling non-specific exceptions and leave the narrowed down ones in explicit except blocks.

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,44 @@ def _flush_unlocked(self):
+                raise
+            else:
+                # The 'try' block completed without an exception
+                if n == 0:
```
So https://docs.python.org/3/library/io.html#io.RawIOBase.write says

> `None` is returned if the raw stream is set not to block and no single byte could be readily written to it. The caller may release or mutate *b* after this method returns, so the implementation should only access *b* during the method call.

Now should this be handled? With a retry? Also, is the difference between `0` and `None` significant?

webknjaz commented on cheroot/errors.py:
```diff
@@ -3,6 +3,16 @@
+
+
+try:
+    from OpenSSL.SSL import SysCallError as _OpenSSL_SysCallError
```
Is it possible to handle this in `cheroot.ssl.pyopenssl` and re-raise as `ConnectionError` or similar so that the external code would only deal with the unified exception and wouldn't need to deal with an import scoped to an optional module?

I think this is an abstraction leak and if we can modify methods in TLS readers/writers to handle their specific cases and re-raise as some normalized `ConnectionError`, then the outer layers wouldn't need to know of such implementation details.

It looks like `NoSSLError` and `FatalSSLAlert` exceptions in this module are trying to achieve this. New refactoring should attempt the same.

webknjaz commented on cheroot/test/test_makefile.py:
```diff
@@ -1,10 +1,12 @@
 
 class MockSocket:
-    """A mock socket."""
+    """A mock socket for emulating buffered I/O."""
```
I didn't realize. Okay, keep this change. But I think `test_flush_unlocked_handles_blocking_io_error` belongs in this test module. I was trying to keep a mapping of `module.py <-> test_module.py` in the test dir. It seems you're mostly testing behavior of makefile in there anyway.

julianz- replied:
i had to move some stuff out of test_makefile.py as one of the linters complained about too many modules?

julianz- replied:
good to know thanks!

webknjaz commented on cheroot/test/test_errors.py:
```diff
@@ -41,3 +45,218 @@ def test_plat_specific_errors(err_names, err_nums):
+UNACCEPTABLE_ERR_NUM = 99999
+
+# Combine known and unknown cases for both types of exceptions
+TEST_CASES = (
```
FTR, I'd expect to use explicit test IDs for these.

webknjaz commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -269,6 +271,41 @@ def __init__(self, *args):
+                ) from err
+
+            # If it's another SysCallError, let it propagate
+            raise
```
Let's fully map the specific exceptions where possible and always convert the exception without reraising so that the outer layer always knows how to handle it w/o the internal exception leaking up the stack.

```suggestion
        except SSL.SysCallError as ssl_syscall_err:
            connection_error_map = {
                errno.EBADF: errors.CherootConnectionError,  # socket is gone?
                errno.ECONNABORTED: errors.CherootConnectionAbortedError,
                errno.ECONNREFUSED: errors.CherootConnectionRefusedError,
                errno.ECONNRESET: errors.CherootConnectionResetError,
                errno.ENOTCONN: errors.CherootConnectionError,
                errno.EPIPE: errors.CherootBrokenPipeError,
                errno.ESHUTDOWN: errors.CherootBrokenPipeError,
            }

            error_code = ssl_syscall_err.args[0] if ssl_syscall_err.args else None
            error_msg = os.strerror(error_code) if error_code is not None else repr(ssl_syscall_err)

            conn_err_cls = connection_error_map.get(error_code, errors.CherootConnectionError)

            raise conn_err_cls(
                error_code,
                f'Faied to {method_name!s} the PyOpenSSL connection: {error_msg!s}',
            ) from ssl_syscall_err
```

julianz- commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,45 @@ def _flush_unlocked(self):
+        and ignored, as these indicate a normal, expected connection teardown.
+        Other exceptions are re-raised.
+        """
+        if self.closed:  # pylint: disable=using-constant-test
```
I added comment

webknjaz commented on cheroot/errors.py:
```diff
@@ -87,4 +93,23 @@ def plat_specific_errors(*errnames):
 
-acceptable_sock_shutdown_exceptions = (BrokenPipeError, ConnectionResetError)
+
+def _is_acceptable_socket_shutdown_error(sock_err):
```
Looks like with the exception conversion this function is becoming redundant.

webknjaz commented on cheroot/server.py:
```diff
@@ -1192,9 +1192,12 @@ def write(self, chunk):
-            self.conn.wfile.write(chunk)
+            data = chunk
+
+        with contextlib.suppress(ConnectionError):
```
Looking deeper, I see that `_process_connections_until_interrupted()` in thread pool is already set up to process connection errors. So this can just be dropped.

If an network error occurrs anytime while processing a single request, there's nothing to do but just log it. And such things should be happening on the top level. Let the connection errors bubble up all the way to that layer and don't suppress them.

The only thing we need to make sure of is that any network errors that present themselves in other ways, are turned into one of the specific connection error exceptions.

webknjaz commented on cheroot/server.py:
I think I saw a few places in this module where `socket.error` / `OSError` is being suppressed. I haven't checked them all but it seems they need to be unshielded too so that the connection error bubbles to the top properly.

julianz- replied:
There's no changes to server.py now.

webknjaz commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -269,6 +271,54 @@ def __init__(self, *args):
+            )
+            conn_err_cls = connection_error_map.get(
+                error_code,
+                errors.CherootConnectionError,
```
Missed this place.

julianz- commented:
Thank you @webknjaz for all your great suggestions and points. I will take a look at everything. The layering is actually more complicated than I had realized.

webknjaz commented:
> The layering is actually more complicated than I had realized.

Yep, that's why I have incomplete pieces of code somewhere locally that I never finalized. It does require some effort..

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,44 @@ def _flush_unlocked(self):
+        try:
+            super().close()
+        except _errors.acceptable_sock_shutdown_exceptions:
+            return
```
This is redundant because `_errors.acceptable_sock_shutdown_exceptions` only contains subclasses of `ConnectionError`.
```suggestion
```

webknjaz commented on cheroot/ssl/pyopenssl.py:
Will this interferre with the newly added decorator/methods? `proxy_wrapper()` below does effectively what we did with `@_morph_syscall_to_connection_error` and so `shutdown()` will end up double-decorated/locked if I'm reading this correctly.

webknjaz commented on cheroot/test/test_makefile.py:
```diff
@@ -51,3 +57,116 @@ def test_bytes_written():
+    try:
+        mock_buffer_writer._flush_unlocked()
+    except Exception as exc:
+        pytest.fail(f'Unexpected exception raised: {type(exc).__name__}')
```
It's unclear why this needs to be wrapped with try/except: tests are supposed to be as simple as possible — adding more logic makes it difficult to reason about them. Additionally, any exception would bubble up and render in the output with the entire stack and respective context attached unlike a short failure message with the stack being lost.

julianz- replied:
This is gone from the current version

webknjaz commented on cheroot/errors.py:
```diff
@@ -66,19 +68,23 @@ def plat_specific_errors(*errnames):
     errno.EPIPE,
     errno.ESHUTDOWN,  # corresponds to BrokenPipeError in Python 3
     errno.ECONNRESET,  # corresponds to ConnectionResetError in Python 3
+    *((errno.WSAENOTSOCK,) if _compat.IS_WINDOWS else ()),
```
@julianz- missed this question?

julianz- replied:
Sorry missed these from October! Now addressed.

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,42 @@ def _flush_unlocked(self):
+        except ConnectionError:
+            return
+        except OSError as err:
+            # Handle EBADF and other acceptable socket shutdown errors
+            if err.errno in _errors.acceptable_sock_shutdown_error_codes:
+                return
```
@julianz- do you know if error codes listed there contain things that aren't represented by `ConnectionError`? I'm wondering if newer Python versions would ever hit this code branch — is it possible that the above `ConnectionError` will handle this?

webknjaz replied:
Also, do we know what's causing the tests to always hit this line?
https://app.codecov.io/gh/cherrypy/cheroot/pull/779?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=cherrypy#d99a7689c2159eab73db39f9e5b89a1d-R70.

It's best to make sure that this isn't due to some mocks but a real simulated scenario.

julianz- replied:
EPIPE/ESHUTDOWN -> BrokenPipeError and ECONNRESET -> ConnectionResetError are ConnectionError subclasses. But EBADF and ENOTCONN are plain OSErrors, so the `except OSError` branch is needed for those. I added test_close_error_handling[ebadf] to cover this path.

Regarding the coverage, the Codecov report is for an older commit — the code has been significantly refactored since then. Not sure whether your question is still a concern?

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,42 @@ def _flush_unlocked(self):
+            # Handle EBADF and other acceptable socket shutdown errors
+            if err.errno in _errors.acceptable_sock_shutdown_error_codes:
+                return
+            raise
```
I wish we have proper logging infra here to see what can actually happen. The coverage report shows that this line is never exercised in the tests.

webknjaz replied:
Perhaps, we could have a warning emitted, though..

julianz- replied:
Covered now. Logging would be nice but maybe for another PR?

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,42 @@ def _flush_unlocked(self):
+        try:
+            super().close()
+        except ConnectionError:
+            return
```
Looks like tests never hit this.

julianz- replied:
Stale code i think. All lines in makefile.py covered now.

webknjaz commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,42 @@ def _flush_unlocked(self):
+            if n == 0:
+                # If nothing was written we need to break
+                # to avoid infinte loops
+                break
```
The tests never hit this line. We should come up with a new regression test.

julianz- replied:
ditto

julianz- commented on cheroot/ssl/pyopenssl.py:
Added tests and pyopenssl was updated in #800 and #801.

julianz- commented on cheroot/makefile.py:
```diff
@@ -32,8 +34,45 @@
+            if not n:
+                # Defensive: write() returned None or other falsy value,
+                # which shouldn't happen but could cause an infinite loop.
+                break
```
I think CodeQL is assuming n is an int but it's not clear that it has to be.

julianz- committed:
Remove socket unraisable exception suppressions from pytest plugin

Now that BufferedWriter.close() handles socket teardown errors
gracefully, these warnings no longer occur in tests.
